File: Algorithms/131.palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string> > result;
        if (s.empty()) return result;
        
        vector<string> palindromes;
        dfs(s, 0, palindromes, result);
        
        return result;
    }
private:
    void dfs(string s, int pos, vector<string> &palindromes, 
             vector<vector<string> > &ret) {
        
        if (pos == s.size()) {
            ret.push_back(palindromes);
            return;
        }
        
        for (int i = pos + 1; i <= s.size(); ++i) {
            string substr = s.substr(pos, i - pos);
            if (!isPalindrome(substr)) {
                continue;
            }
            
            palindromes.push_back(substr);
            dfs(s, i, palindromes, ret);
            palindromes.pop_back();
        }
    }
    
    bool isPalindrome(string s) {
        if (s.empty()) return false;
        
        int n = s.size();
        for (int i = 0; i < n; ++i) {
            if (s[i] != s[n - i - 1]) return false;
        }
        
        return true;
    }
};

File: Algorithms/131.palindrome-partitioning/palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<List<String>>();
        if (s == null || s.isEmpty()) return result;
        
        List<String> palindromes = new ArrayList<String>();
        dfs(s, 0, palindromes, result);
        
        return result;
    }
    private void dfs(String s, int pos, List<String> palindromes, 
                     List<List<String>> ret) {

        if (pos == s.length()) {
            ret.add(new ArrayList<String>(palindromes));
            return;
        }
        
        for (int i = pos + 1; i <= s.length(); i++) {
            String substr = s.substring(pos, i);
            if (!isPalindrome(substr)) {
                continue;
            }
            
            palindromes.add(substr);
            dfs(s, i, palindromes, ret);
            palindromes.remove(palindromes.size() - 1);
        }
    }
    
    private boolean isPalindrome(String s) {
        if (s == null || s.isEmpty()) return false;
        
        int n = s.length();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) != s.charAt(n - i - 1)) return false;
        }
        
        return true;
    }
}
